src/lib.rs

```diff
@@ -0,0 +1,53 @@
+use ext_php_rs::binary::Binary;
+use ext_php_rs::prelude::*;
+
+/// Creates a scrypt password hash.
+///
+/// @param string password The clear text password
+/// @param string salt     The salt to use for the hash
+/// @param u8     cpu_difficulty        The CPU difficulty [default=32768]
+/// @param u32    memory_difficulty        The memory difficulty [default=8]
+/// @param u32    parallel_difficulty        The parallel difficulty [default=1]
+/// @param u23    len      The length of the generated hash [default=8]
+///
+/// @return string The hashed password
+#[php_function]
+pub fn scrypt(
+  password: &str,
+  salt: Binary<u8>,
+  cpu_difficulty: Option<u32>,
+  memory_difficulty: Option<u32>,
+  parallel_difficulty: Option<u32>,
+  len: Option<usize>,
+) -> Result<String, String> {
+  let password = password.as_bytes();
+
+  let n = match cpu_difficulty {
+    Some(data) => fast_math::log2(data as f32) as u8,
```

Comment:
Can't we use `f64::log2()` here?

So having the `cpu_difficulty` be `Option<f64>` and do `data.log2()` here?

Reply:
Reference => https://doc.rust-lang.org/std/primitive.f64.html#method.log2